src/cargo/util/config.rs

```diff
@@ -21,6 +22,30 @@ use util::toml as cargo_toml;
 
 use self::ConfigValue as CV;
 
+#[derive(PartialEq, Eq, Hash)]
+struct ConfigFile {
+    path: Option<PathBuf>,
+}
+
+impl ConfigFile {
+    pub fn new(path: PathBuf) -> ConfigFile {
+        let canonical = match fs::canonicalize(path) {
```

Comment:
Now that I'm thinking, if I remove the `fs::canonicalize` call I don't really need the struct. I put it only to simplify the "canonicalization" process. I think it would be simpler if I remove the struct and leave only the `HashSet`.